wolfcrypt/src/port/Renesas/renesas_sce_ra6m3g.c

```diff
@@ -0,0 +1,1146 @@
+        else if (op == AES_SCE_DECRYPT && keySize == AES256_KEYSIZE)
+            ret = HW_SCE_AES_256EcbDecrypt((const uint32_t*) aes->key,
+                                           (const uint32_t)  num_words,
+                                           (const uint32_t*) in,
+                                           (uint32_t*)       out);
+#endif
+        else
+            ret = BAD_FUNC_ARG;
+        wolfSSL_CryptHwMutexUnLock();
+    }
+
+    if (ret != FSP_SUCCESS && ret != BAD_FUNC_ARG) {
+        WOLFSSL_MSG("Hardware error.");
+        ret = WC_HW_E;
+    }
+    return ret;
+}
+
+/*
+   AES-CTR Encrypt with Renesas RA Hardware
+
+   Inputs:
+        aes Contains AES key and parameters for stream re-use
+        in  Plaintext to encrypt
+        sz  Size of plaintext
+   Outputs:
+        out Ciphertext message
+
+   return  FSP_SUCCESS:     HW completed successfully
+           BAD_FUNC_ARG:    Invalid Argument
+           WC_HW_E:         Hardware could not encrypt message.
+*/
+int wc_Renesas_AesCtrEncrypt(Aes* aes, byte* out, const byte* in, word32 sz) {
+    word32 keySize = 0;
+    int ret = WOLFSSL_SUCCESS;
+    const byte* tmp;
+    uint32_t* outTmp;
+    byte inTmp[AES_BLOCK_SIZE] = {0};
+
+    WOLFSSL_ENTER("wc_Renesas_AesCtrEncrypt");
+
+    if (aes == NULL || out == NULL || in == NULL)
+        ret = BAD_FUNC_ARG;
+
+    wc_AesGetKeySize(aes, &keySize);
+    if (!(keySize == AES128_KEYSIZE || keySize == AES256_KEYSIZE))
+        ret = BAD_FUNC_ARG;
+
+    /* Use remaining AES stream from previous non-AES_BLOCK_SIZE operation */
+    tmp = (byte*)aes->tmp + AES_BLOCK_SIZE - aes->left;
+    while (aes->left && sz && ret != BAD_FUNC_ARG) {
+       *(out++) = *(in++) ^ *(tmp++);
+       aes->left--;
+       sz--;
+    }
+
+    while (sz) {
+        tmp = in;
+        outTmp = (uint32_t*) out;
+
+        if (sz < AES_BLOCK_SIZE) {
+            /* Copy remaining bytes into AES_BLOCK_SIZE buffer */
+            XMEMCPY(inTmp, in, sz);
+            tmp = inTmp;
+            /* Set output to aes->tmp for future stream re-use */
+            outTmp = (uint32_t*) aes->tmp;
+        }
+
+        if (wolfSSL_CryptHwMutexLock() != 0)
+            break;
+
+        if (keySize == AES128_KEYSIZE)
+            ret = HW_SCE_AES_128CtrEncrypt((const uint32_t*) aes->key,
+                                           (const uint32_t*) aes->reg,
+                                           (const uint32_t)  AES_BLOCK_WORDS,
+                                           (const uint32_t*) tmp,
+                                           (uint32_t*)       outTmp,
+                                           (uint32_t*)       aes->reg);
+        else if (keySize == AES256_KEYSIZE)
+            ret = HW_SCE_AES_256CtrEncrypt((const uint32_t*) aes->key,
+                                           (const uint32_t*) aes->reg,
+                                           (const uint32_t)  AES_BLOCK_WORDS,
```

Comment:
I left a comment about that. I didnt want to change.  I tried that at first.  I may have implemented that wrong with the stream re-use, but it might've been a HW or HW driver error.  A lot of their API are very particular about inputs being block sizes.